### PR TITLE
fix #22626: automatic system dividers

### LINKED ISSUE #22626: Horizontal frames vis à vis instrument names
### Issue type

Engraving bug (incorrect score rendering)

### Bug description

Often, I would indent a system by creating an horizontal frame--this would display in Ms3.x as:
(Margin)--Frame--Instrument name--bracket/system start

In Ms4, this displays as:
(Margin)--Instrument name, (overlaid with) Frame--Bracket/system start

This creates a huge gap between the instrument name and the bracket. Deleting the horizontal frame and re-creating it does not help. I found I could move the instrument names individually using CTRL->, although I would have to offset the frames to move them out of the way, then return them to position. However, I find that when I close the file and reopen it, the instrument names are back against the margin--even if I have hit CTRL-S before closing.

image

before m.17 and m.21 in the attached file.
[Asperges_Me.4.pdf](https://github.com/musescore/MuseScore/files/15151797/Asperges_Me.4.pdf)

### MuseScore Version

4.2.1

### Regression

Yes, this used to work in MuseScore 3.x and now is broken

### Operating system

Windows 11

### PR DESCRIPTION
**Pull request by MarcSabatella**
This is a proposed new feature as per popular request.  It works by adding settings to Style / General / System that control automatic generation of dividers.  When enabled, they are placed between systems on each page, in Score::layoutPage(), using the symbol and offsets specified in the style settings.  They are added to the element list for the measure, just like layout breaks.  The dividers are marked as "generated" and not written to the file unless you modify them (eg, via the Inspector).  For instance, it works nicely to turn on the dividers, then mark some invisible and change the horiztonal / vertical offset of others, and this is preserved on reload.

When implementing this, I chose to use ordinary Symbol elements, and just do a little special casing as necessary: symbols with Measures as parents are treated as dividers.  This was based on how we handle measure numbers - ordinary text with just a small amount of special casing.  However, as I implemented the style options, it became more clear that I should probably have a new element type.  Assuming others agree, I'll make that change, but I figured I'd wait to get feedback on the basic design first.  And of course, once it's all finalized, I'll add tests :-)

The implementation adds new style options and, more significantly, a new tag "SystemDivider" that is written to the file, so this should also come with a version number bump.  I haven't done that yet either.

### CODE REVIEW

**Comment by MarcSabatella** [comment]
I went ahead and added the new element type, SystemDivider.  I don't think I've added a new element type before, so I'm not sure I got all the details right, but things do seem to work.

File: libmscore/system.cpp [hunk 1]
```diff
@@ -39,6 +39,8 @@
 #include "chordrest.h"
 #include "iname.h"
 #include "spanner.h"
+#include "sym.h"
+#include "symbol.h"
```
Comment (lasconic) [inline libmscore/system.cpp]: `#include "symbol.h"` is probably not needed since it's already included in system.h ?

File: mscore/editstyle.ui [hunk 2]
```diff
@@ -6,7 +6,7 @@
    <rect>
     <x>0</x>
     <y>0</y>
-    <width>923</width>
+    <width>1150</width>
```
Comment (lasconic) [inline mscore/editstyle.ui]: Is this really needed or an artifact of QtDesigner?

Comment (MarcSabatella) [inline mscore/editstyle.ui]: Artifact.  I believe we override this in the code anyhow, to get rid of the constant issues we were having with this dialog?

File: libmscore/system.cpp (hunk 1 above)
Comment (MarcSabatella) [inline libmscore/system.cpp]: Could be.  Seems I wouldn't have added it if I hadn't gotten a compilation error, but maybe I was confused.

File: mscore/editstyle.ui (hunk 2 above)
Comment (lasconic) [inline mscore/editstyle.ui]: Yes. That's fine.

File: libmscore/system.cpp (hunk 1 above)
Comment (lasconic) [inline libmscore/system.cpp]: #include "symbol.h" shouldn't be needed. Could you remove it?

**Comment by lasconic** [comment]
I tried this PR a bit more. I expected the dividers to be positioned in between two systems at the same distance of the two systems even when using spacers and so on. The default placement is quite important because any offset added by the users in style or for each divider will be applied on top of this placement. Any thought?

**Comment by MarcSabatella** [comment]
Divders are centered vertically between systems now.

**Pull request merged**